[Bug]: aria_snapshot generates invalid yaml when combined with an aria-label attribute

### Version

1.49.0

### Steps to reproduce

1. Save the below dockerfile
2. Build and run with `docker run $(docker build -q .)`
3. See the error, along with the generated yaml in stdout 

```dockerfile
FROM mcr.microsoft.com/playwright/python:v1.49.0

RUN pip3 install pytest pytest-playwright

RUN mkdir /test && \
    # create html file
    echo '<!doctype html>' >> /test/index.html && \
    echo '<html>' >> /test/index.html && \
    echo '    <body>' >> /test/index.html && \
    echo '        <h1 aria-label="@header - label">@header</h1>' >> /test/index.html && \
    echo '    </body>' >> /test/index.html && \
    echo '</html>' >> /test/index.html && \
    # create python file
    echo 'from playwright.sync_api import Page, expect' >> /test/test_main.py && \
    echo 'def test_main(page: Page):' >> /test/test_main.py && \
    echo '    page.goto("file:///test/index.html")' >> /test/test_main.py && \
    echo '    print(page.locator("body").aria_snapshot())' >> /test/test_main.py && \
    echo '    expect(page.locator("body")).to_match_aria_snapshot(' >> /test/test_main.py && \
    echo "        '- heading \"@header - label\" [level=1]: @header'" >> /test/test_main.py && \
    echo '    )' >> /test/test_main.py

CMD pytest /test --browser chromium --tracing retain-on-failure --output /output -rP
```

### Expected behavior

Valid yaml should be generated:

```yaml
- heading "@header - label" [level=1]: "@header"
```

### Actual behavior

Invalid yaml is generated:

```yaml
- heading "@header - label" [level=1]: @header
```

### Additional context

The problem seems to be elements that have text that starts with "@" and have an `aria-label` attribute:

```html
<!doctype html>
<html>
    <body>
        <h1 aria-label="@header - label">@header</h1>
    </body>
</html>
```

pytest output:

```
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-8.3.3, pluggy-1.5.0
rootdir: /test
plugins: playwright-0.6.2, base-url-2.1.0
collected 1 item

test/test_main.py F                                                      [100%]

=================================== FAILURES ===================================
_____________________________ test_main[chromium] ______________________________

page = <Page url='file:///test/index.html'>

    def test_main(page: Page):
        page.goto("file:///test/index.html")
        print(page.locator("body").aria_snapshot())
>       expect(page.locator("body")).to_match_aria_snapshot(
            '- heading "@header - label" [level=1]: @header'
        )

test/test_main.py:5: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/local/lib/python3.12/dist-packages/playwright/_impl/_locator.py:712: in _expect
    result = await self._frame._channel.send_return_as_dict(
usr/local/lib/python3.12/dist-packages/playwright/_impl/_connection.py:67: in send_return_as_dict
    return await self._connection.wrap_api_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <playwright._impl._connection.Connection object at 0x73211b126060>
cb = <function Channel.send_return_as_dict.<locals>.<lambda> at 0x73211b133d80>
is_internal = False

    async def wrap_api_call(
        self, cb: Callable[[], Any], is_internal: bool = False
    ) -> Any:
        if self._api_zone.get():
            return await cb()
        task = asyncio.current_task(self._loop)
        st: List[inspect.FrameInfo] = getattr(task, "__pw_stack__", inspect.stack())
        parsed_st = _extract_stack_trace_information_from_stack(st, is_internal)
        self._api_zone.set(parsed_st)
        try:
            return await cb()
        except Exception as error:
>           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
E           playwright._impl._errors.Error: LocatorAssertions.to_match_aria_snapshot: Plain value cannot start with reserved character @ at line 1, column 40:
E           
E           - heading "@header - label" [level=1]: @header
E                                                  ^

usr/local/lib/python3.12/dist-packages/playwright/_impl/_connection.py:528: Error
----------------------------- Captured stdout call -----------------------------
- heading "@header - label" [level=1]: @header
============================== 1 failed in 0.94s ===============================
```

### Environment

```Text
- Operating System: elementary OS 7.1 (based on Ubuntu 22.04.5 LTS)
- CPU: x86_64
- Browser: Chromium
- Python Version: 3.12
```

--- a/packages/playwright-core/src/server/injected/yaml.ts
+++ b/packages/playwright-core/src/server/injected/yaml.ts
@@ -62,12 +62,8 @@ function yamlStringNeedsQuotes(str: string): boolean {
   if (/^-\s/.test(str))
     return true;
 
-  // Strings that start with a special indicator character need quotes
-  if (/^[&*\],].*/.test(str))
-    return true;
-
-  // Strings containing ':' followed by a space or at the end need quotes
-  if (/:(\s|$)/.test(str))
+  // Strings containing ':' or '\n' followed by a space or at the end need quotes
+  if (/[\n:](\s|$)/.test(str))
     return true;
 
   // Strings containing '#' preceded by a space need quotes (comment indicator)
@@ -78,20 +74,16 @@ function yamlStringNeedsQuotes(str: string): boolean {
   if (/[\n\r]/.test(str))
     return true;
 
-  // Strings starting with '?' or '!' (directives) need quotes
-  if (/^[?!]/.test(str))
-    return true;
-
-  // Strings starting with '>' or '|' (block scalar indicators) need quotes
-  if (/^[>|]/.test(str))
-    return true;
-
-  // Strings starting with quotes need quotes
-  if (/^["']/.test(str))
+  // Strings starting with indicator characters or quotes need quotes
+  if (/^[&*\],?!>|@"'#%]/.test(str))
     return true;
 
   // Strings containing special characters that could cause ambiguity
   if (/[{}`]/.test(str))
+    return true;
+
+  // Non-string types recognized by YAML
+  if (!isNaN(Number(str)) || ['y', 'n', 'yes', 'no', 'true', 'false', 'on', 'off', 'null'].includes(str.toLowerCase()))
     return true;
 
   return false;

--- a/tests/page/page-aria-snapshot.spec.ts
+++ b/tests/page/page-aria-snapshot.spec.ts
@@ -508,3 +508,57 @@ it('should handle long strings', async ({ page }) => {
       - region: ${s}
   `);
 });
+
+it('should escape special yaml characters', async ({ page }) => {
+  await page.setContent(`
+    <a href="#">@hello</a>@hello
+    <a href="#">]hello</a>]hello
+    <a href="#">hello\n</a>
+    hello\n<a href="#">\n hello</a>\n hello
+    <a href="#">#hello</a>#hello
+  `);
+
+  await checkAndMatchSnapshot(page.locator('body'), `
+    - link "@hello"
+    - text: "@hello"
+    - link "]hello"
+    - text: "]hello"
+    - link "hello"
+    - text: hello
+    - link "hello"
+    - text: hello
+    - link "#hello"
+    - text: "#hello"
+  `);
+});
+
+it('should escape special yaml values', async ({ page }) => {
+  await page.setContent(`
+    <a href="#">true</a>False
+    <a href="#">NO</a>yes
+    <a href="#">y</a>N
+    <a href="#">on</a>Off
+    <a href="#">null</a>NULL
+    <a href="#">123</a>123
+    <a href="#">-1.2</a>-1.2
+    <input type=text value="555">
+  `);
+
+  await checkAndMatchSnapshot(page.locator('body'), `
+    - link "true"
+    - text: "False"
+    - link "NO"
+    - text: "yes"
+    - link "y"
+    - text: "N"
+    - link "on"
+    - text: "Off"
+    - link "null"
+    - text: "NULL"
+    - link "123"
+    - text: "123"
+    - link "-1.2"
+    - text: "-1.2"
+    - textbox: "555"
+  `);
+});

